token-metadata/program/src/instruction.rs

```diff
@@ -0,0 +1,120 @@
+use {
+    borsh::{BorshDeserialize, BorshSerialize},
+    solana_program::{
+        instruction::{AccountMeta, Instruction},
+        pubkey::Pubkey,
+        sysvar,
+    },
+    std::str::FromStr,
+};
+
+#[repr(C)]
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug, Clone)]
+/// Args for Create call
+pub struct CreateMetadataAccountArgs {
+    /// The name of the asset
+    pub name: String,
+    /// The symbol for the asset, ie, AAPL or SHOES
+    pub symbol: String,
+    /// URI pointing to JSON representing the asset
+    pub uri: String,
+}
+
+#[repr(C)]
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug, Clone)]
+/// Args for init call
+pub struct InitMetadataAccountArgs {
+    /// The name of the asset
+    pub name: String,
+    /// The symbol for the asset, ie, AAPL or SHOES
+    pub symbol: String,
+    /// URI pointing to JSON representing the asset
+    pub uri: String,
+}
+
+#[repr(C)]
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug, Clone)]
+/// Args for update call
+pub struct UpdateMetadataAccountArgs {
+    /// URI pointing to JSON representing the asset
+    pub uri: String,
+}
+
+/// Instructions supported by the Metadata program.
+#[derive(BorshSerialize, BorshDeserialize, Clone)]
+pub enum MetadataInstruction {
+    /// Create an  Owner and  Metadata objects.
+    ///   0. `[writable]`  Owner key (pda of ['metadata', program id, name, symbol])
+    ///   1. `[writable]`  metadata key (pda of ['metadata', program id, mint id])
+    ///   2. `[]` Mint of token asset
+    ///   3. `[signer]` Mint authority
+    ///   4. `[signer]` payer
+    ///   5. `[]` System program
+    CreateMetadataAccounts(CreateMetadataAccountArgs),
+
+    /// Update an  Metadata (name/symbol are unchangeable)
+    ///   0. `[writable]`  Metadata account
+    ///   1. `[signer]` Owner key
+    ///   2. `[]`  Owner account
+    UpdateMetadataAccounts(UpdateMetadataAccountArgs),
+}
```

Comment:
I don't see an instruction to update the owner address?

Reply:
Also we could just default the owner address to the mint authority, since only the mint authority can create the metadata to begin with.   Then the "SetUpdateAuthority" instruction only needs to be invoked if the mint authority wants to delegate update authority to another address